### ISSUE TITLE
TrayNotification <missing message content>

### ISSUE THREAD

**Issue by philippsander** [NativeScript/push-plugin#98]
I copied the sample code and actually recieve a message.

`PushPlugin.onMessageReceived(function callback(message, data, notification) {
    });`

However, my first parameter is null, the second one contains the data I actually need. 

How do I manipulate Notification (message and action)?

I'm using 0.1.2

**Comment by philippsander** [comment]
can someone help me with this please?

**Comment by AntonDobrev** [comment]
@philippsander At some point and depending on your payload, the `message` parameter will be empty. Please read more here - https://github.com/NativeScript/push-plugin#receive-and-handle-messages-from-fcm-on-android 

If there is further ambiguity or uncertainty let me know the payload you are sending and I will try to help you out. 

Happy coding!

**Comment by philippsander** [comment]
sorry... I don't understand how to use it.

`PushPlugin.onMessageReceived(function callback(message, data, notification) {
      
    });`

this is my method. How am I supposed to change the message of the notification?

**Comment by AntonDobrev** [comment]
@philippsander 

Philip, here are some more explanations. However, you need first to be examine the payload sent to your app and what you would like to handle. 

For example, payload for Android consists of a JSON object that contains either a `data` or `notification` keys in it, or both. 

Utilizing the callback parameters depends on what you are sending with the payload. The `message` argument is usually precooked for you if there is a `message` key in the `data` payload. For full experience usually you need to leverage the second and third parameter depending on the payload and the app state (more in the documentation link from my previous reply). That is why I asked for your payload. 

Example:
```
{
    "data":{
        "title":"my title",
        "message":"my message", // will be populated in the first argument of the callback 
... other keys
    },
   "notification": {
   // as per the docs https://firebase.google.com/docs/cloud-messaging/concept-options
    }
}
```

Regards, 
Anton

**Comment by philippsander** [comment]
The firebase system is coming from someone else. I need to manipulate the notification in my app if possible. is that possible or not?

**Comment by AntonDobrev** [comment]
@philippsander 

Yes, you can leverage the `data` and `notification` arguments of the callback:

```
PushPlugin.onMessageReceived(function callback(message, data, notification) {

});`

```

**Comment by philippsander** [comment]
I feel so stupid.... how?!

data is a string... message and notification are null

**Comment by philippsander** [comment]
i just want to change the message of the notification from within my app... it can't be that hard.

**Comment by AntonDobrev** [comment]
@philippsander To be honest, initially I did not get right your question.

According to your payload, your app will receive in the second argument of the callback an object with the following keys: `Header`, `Body`, and the plugin set property `foreground`. However, the entry in the tray will note feature any title or message because the plugin is expecting them as "title" and "message" keys in the `data` payload. 

To change the title and message you have the following two ways. Note that you cannot change them from the app because the notification is already processed as of that moment in time (however, you should be able to create with the native APIs a new notification and place it in the tray if you need to). 

1. From the server: 
 - Set the `title` and `message` keys in the `data` payload (the plugin is designed to set them in the tray):
```
{
    "data":{
        "title":"title 1",
        "message":"message 1"
    }
}
```
- set the `title` and `body` of the `notification`:

```
{
    "notification":{
        "title":"my title",
        "body":"my message"
    }
}
```

2. From the plugin native code:

As you can see in the native code [here](https://github.com/NativeScript/push-plugin/blob/master/native-src/android/app/src/main/java/com/telerik/pushplugin/NotificationBuilder.java#L72) the tray's notification is built there with title and message. You can use default values if you need to by modifying the plugin. 

I hope this answers your question. 

Some further notes:

The plugin automatically handles some keys in the data object like `message`, `title`, `color`, `smallIcon`, `largeIcon `and uses them to construct a notification entry in the tray. More information on these keys is available in the documentation of the Telerik Platform Notifications service documentation [article](http://docs.telerik.com/platform/backend-services/javascript/push-notifications/send-and-target/push-send-target-examples).

**Comment by philippsander** [comment]
I think it would be a handy feature to be able to modify the msgData.

beforeMessageReceived for example
